### ISSUE TITLE
sinks stop after Redpanda restart: Producer attempted a transactional operation in an invalid state

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#14098]
### What version of Materialize are you using?

9457d33a4dbae7770260a37489925d868fd869bc

### How did you install Materialize?

Docker image

### What is the issue?

If Redpanda is restarted, the sinks stop producing stuff. The errors in the log can be seen below.

To reproduce:

1. Clone `git@github.com:philip-stoev/materialize.git`, branch `checks-restart-redpanda`
2. Run
```
cd test/platform-checks
 ./mzcompose run default --check=SinkUpsert --scenario=RestartRedpanda
```
3. It will hang as follows:
```
> SELECT l_v, l_k, SUM(c) FROM (   SELECT (after).l_v, (after).l_k, (after).c FROM sink_view1   UNION ALL   SELECT (before).l_v, (before).l_k, -(before).c FROM sink_view1 ) GROUP BY l_v, l_k HAVING SUM(c) > 0;
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s
```

What this test does is create a sink, and then a source on top of the sink so that it can be re-ingested and validated. The query above examines this sink-derived source in question for completeness.

4. Connect to the Mz instance to run the query manually and compare the results against the expected ones from `misc/python/materialize/checks/sink.py`. You will observe that the sink has only produced data until the first Redpanda restart after the sink was created.

5. Log into the `platform-checks-materialized-1` container and use 

 ```
kill -9 `pidof storaged`
```

To kill all the storageds so that they are restarted. Observe now that the validation query hangs altogether rather than producing a wrong/stale/old/incomplete result

6. To confirm that the sink and not the source derived from the sink is the problem, create a new source using the source definition from `SHOW CREATE SOURCE sink_view1` using a different name. Observe that running the validation query against the new source still does not return the correct result.

### Relevant log output

```source
compute-cluster-1-replica-1: 2022-08-05T11:33:42.850983Z  INFO mz_compute::sink::kafka: Retriable error in kafka sink: RDKafkaError(Failed to flush all outstanding messages within the transaction timeout: 1 message(s) remaining)
compute-cluster-1-replica-1: 2022-08-05T11:33:44.163287Z ERROR librdkafka: ADDPARTS [thrd:main]: TxnCoordinator/0: Failed to add partition "sink-sink1-consistency" [0] to transaction: Broker: Producer attempted a transactional operation in an invalid state
compute-cluster-1-replica-1: 2022-08-05T11:33:44.163307Z ERROR librdkafka: TXNERR [thrd:main]: Fatal transaction error: Failed to add partitions to transaction: Broker: Producer attempted a transactional operation in an invalid state (INVALID_TXN_STATE)
compute-cluster-1-replica-1: 2022-08-05T11:33:44.163312Z ERROR librdkafka: FATAL [thrd:main]: Fatal error: Broker: Producer attempted a transactional operation in an invalid state: Failed to add partitions to transaction: Broker: Producer attempted a transactional operation in an invalid state
compute-cluster-1-replica-1: 2022-08-05T11:33:44.163356Z ERROR mz_kafka_util::client: librdkafka: Global error: Fatal (Local: Fatal error): Fatal error: Broker: Producer attempted a transactional operation in an invalid state: Failed to add partitions to transaction: Broker: Producer attempted a transactional operation in an invalid state
compute-cluster-1-replica-1: 2022-08-05T11:33:44.163538Z  INFO mz_compute::sink::kafka: shutting down kafka sink: kafka-u3
```

**Comment by aljoscha** [comment]
Thanks! seems we need to tackle this after @cjubb39 is done moving Sinks over to storage.

**Comment by uce** [comment]
@aljoscha I moved this to P2 since we will only get to it after the sinks move lands (I've also linked it in the respective epic to not loose track).

**Comment by aljoscha** [comment]
many thanks! :bow:

**Comment by cjubb39** [comment]
We expect the above PR to close this out.  This error (`Producer attempted a transactional operation in an invalid state`) may still occur but it'd result in the sink crashing and restarting which should resolve things quickly.  We expect this error to be transient so a restart should fix.

**Comment by philip-stoev** [comment]
Yes but actually no. The issue is immediately repeatable using the original instructions above. I have rebased the branch where the test resides against the latest main.

**Comment by benesch** [comment]
I tried to reproduce this just now on your rebased branch, @philip-stoev, but ... the test succeeded! Is there any nondeterminism in the test? Is it possible I need to run it a few times to see the failure?

I'll try again on my Linux machine, too.

**Comment by benesch** [comment]
Reproduced instantly on my Linux machine. There were several fatal errors that were _not_ causing the sink to crash `storaged`, but rather to silently hang around forever. I pushed up a fix to your PR for the failing test (#14097). Thanks very much for double checking this one, Philip.

**Comment by cjubb39** [comment]
@philip-stoev sorry I wasn't more thorough in trying to repro!  Will add trying repros on a linux machine to the toolset
